### PR TITLE
bpo-33475: Fix and improve converting annotations to strings.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Fix miscellaneous bugs in converting annotations to strings.
* Optimize parenthesis in the string representation.
* Clean up the C code.

https://bugs.python.org/issue33475

### CODE REVIEW

**Comment by gvanrossum** [comment]
@ambv should review it

File: Misc/NEWS.d/next/Core and Builtins/2018-05-13-01-26-18.bpo-33475.rI0y1U.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Fixed miscellaneous bugs in onverting annotations to strings and optimize
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Core and Builtins/2018-05-13-01-26-18.bpo-33475.rI0y1U.rst]: converting, optimized (to make tense consistent), parentheses (on the next line)

**Comment by ambv** [comment]
Thanks for the PR, @serhiy-storchaka! I won't lie, I don't like the macros, I find them hiding actual control flow in the functions. But I won't argue with all the fixes you made.

**Comment by ambv** [comment]
Oops, looks like 3.7 needs help.

**Comment by serhiy-storchaka** [comment]
Thank you for merging the PR and sorry for using macros. My reasons are that they make the code clearer to me they are already used everywhere in the AST processing. In any case this patch changed virtually every line of the code. Macros can help to avoid such large changes in future (if we will add a hidden parameter passed to most functions) or can help to automate changes because they uses are one-line and well searchable.

**Pull request merged**